File: Applied Course/4.Problem Solving/5.Problems on Trees/15.Convert a given Binary Tree to Doubly Linked List.py
class node: 

    def __init__(self, info): 
        self.info = info  
        self.left = None
        self.right = None 

def BtoDLLUtil(root):
    
    if root.left:
        left = BtoDLLUtil(root.left) 
        while left.right: 
            left = left.right 
  
        left.right = root 
        root.left = left 

    if root.right:
        right = BtoDLLUtil(root.right) 
        while right.left: 
            right = right.left 
 
        right.left = root 
        root.right = right 
  
    return root 

def BtoDLL(root):
    if root==None:
        return root
    
    root=BtoDLLUtil(root)
    
    while root.left:
        root=root.left
    return root

def printList(root):
    if root==None:
        return
    
    while(root):
        print(root.info) 
        root = root.right

if __name__=='__main__':
    root=None
    root=node(10)
    root.left=node(20)
    root.right=node(30)
    root.left.left=node(40)
    root.left.right=node(50)
    root.right.left=node(60)
    root.right.right=node(70)
    root=BtoDLL(root)
    printList(root)

File: Applied Course/4.Problem Solving/5.Problems on Trees/15.Convert a given Binary Tree to Doubly Linked List.c
#include <stdio.h> 
#include <stdlib.h> 
  
// Structure for tree and linked list 
struct node
{ 
    int data; 
    struct node *left, *right; 
}; 
  

void BinaryToDLL(struct node* root, struct node** head) 
{ 
    // Base cases 
    if (!root) 
        return; 
  
    // Recursively convert right subtree 
    BinaryToDLL(root->right, head); 
  
    // insert root into DLL 
    root->right = *head; 
  
    // Change left pointer of previous head 
    if (*head) 
        (*head)->left = root; 
  
    // Change head of Doubly linked list 
    *head = root; 
  
    // Recursively convert left subtree 
    BinaryToDLL(root->left, head); 
} 
  
// create a new node 
struct node* newNode(int data) 
{ 
    struct node* temp = (struct node*)malloc(sizeof(struct node)); 
    temp->data = data; 
    temp->left = temp->right = NULL; 
    return temp; 
} 
  
void printList(struct node* head) 
{ 
     
    while (head) 
    { 
        printf("%d ", head->data); 
        head = head->right; 
    } 
} 
  
// Driver program to test above function 
int main() 
{ 
    struct node* root = newNode(5); 
    root->left = newNode(3); 
    root->right = newNode(6); 
    root->left->left = newNode(1); 
    root->left->right = newNode(4); 
    root->right->right = newNode(8); 
    root->left->left->left = newNode(0); 
    root->left->left->right = newNode(2); 
    root->right->right->left = newNode(7); 
    root->right->right->right = newNode(9); 
  
    struct node* head = NULL; 
    BinaryToDLL(root, &head); 
  
    printList(head); 
  
    return 0; 
} 